File: MCRRTImp/src/main/java/lab/mars/MCRRTImp/Vector2BasedImp/EyeSight.java
package lab.mars.MCRRTImp.Vector2BasedImp;

import lab.mars.MCRRTImp.Vector2BasedImp.Vector2;
import lab.mars.RRTBase.Obstacle;

public class EyeSight implements Obstacle<Vector2> {
    private Vector2 origin;

    private Vector2 direction;

    private double radius;

    private double angle;

    public EyeSight(Vector2 origin, Vector2 direction, double radius, double angle) {
        this.origin = origin;
        this.direction = direction;
        this.radius = radius;
        this.angle = angle;
    }

    @Override
    public boolean contains(Vector2 o) {
        return !((this.origin.distance2(o) <= this.radius * this.radius) && direction.angle(o.cpy().subtract(origin)) <= this.angle / 2.0);
    }
}

File: MCRRTImp/src/test/kotlin/infrastructure/Vector2BasedImp/EyeSight.kt
package infrastructure.Vector2BasedImp

import lab.mars.MCRRTImp.base.Obstacle

class EyeSight(private val origin: Vector2, private val direction: Vector2, private val radius: Double, private val angle: Double) : Obstacle<Vector2> {

    override operator fun contains(o: Vector2): Boolean {
        return !(this.origin.distance2(o) <= this.radius * this.radius && direction.angle(o.cpy().subtract(origin)) <= this.angle / 2.0)
    }
}
